core/src/main/java/org/chodavarapu/datamill/http/Client.java

```diff
@@ -61,108 +75,155 @@ protected URLConnection createConnection(String uri) throws IOException {
+                if (options != null && options.size() > 0) {
+                    Object connectTimeout = options.get(Request.OPTION_CONNECT_TIMEOUT);
+                    if (connectTimeout instanceof Integer) {
+                        RequestConfig requestConfig = RequestConfig.custom()
+                                .setConnectTimeout((int) connectTimeout)
+                                .build();
+                        ((HttpRequestBase) request).setConfig(requestConfig);
+                    }
                 }
-            }
 
-            if (headers != null) {
-                for (Map.Entry<String, String> header : headers.entries()) {
-                    httpConnection.addRequestProperty(header.getKey(), header.getValue());
+                if (headers != null) {
+                    for (Map.Entry<String, String> header : headers.entries()) {
+                        request.addHeader(header.getKey(), header.getValue());
+                    }
                 }
-            }
 
-            if (entity != null) {
-                writeEntityOutOverConnection(entity, httpConnection);
-            }
+                if (entity != null) {
+                    if (!(request instanceof HttpEntityEnclosingRequestBase)) {
+                        throw new IllegalArgumentException("Expecting to write an entity for a request type that does not support it!");
+                    }
+
+                    pipedOutputStream = buildPipedOutputStream();
+                    pipedInputStream = buildPipedInputStream();
+                    pipedInputStream.connect(pipedOutputStream);
+
+                    BasicHttpEntity httpEntity = new BasicHttpEntity();
+                    httpEntity.setContent(pipedInputStream);
+                    ((HttpEntityEnclosingRequestBase) request).setEntity(httpEntity);
 
-            logger.debug("Making HTTP request {} {}", method.name(), composedUri);
-            if (headers != null && logger.isDebugEnabled()) {
-                logger.debug("  HTTP request headers:");
-                for (Map.Entry<String, String> header : headers.entries()) {
-                    logger.debug("    {}: {}", header.getKey(), header.getValue());
+                    writeEntityOutOverConnection(entity, pipedOutputStream);
                 }
-            }
 
-            int responseCode = httpConnection.getResponseCode();
-            InputStream inputStream = httpConnection.getInputStream();
+                logger.debug("Making HTTP request {} {}", method.name(), composedUri);
+                if (headers != null && logger.isDebugEnabled()) {
+                    logger.debug("  HTTP request headers:");
+                    for (Map.Entry<String, String> header : headers.entries()) {
+                        logger.debug("    {}: {}", header.getKey(), header.getValue());
+                    }
+                }
+
+                CloseableHttpResponse httpResponse = httpclient.execute(request);
```

Comment:
I think because we wrap the entity for this itno a stream, it will get closed but should check whether this needs to be closed explicitly